src/order/ideal.lean

```diff
@@ -184,15 +163,47 @@ lemma ideal_Inter_nonempty_iff :
   ideal_Inter_nonempty P ↔ ∃ a : P, ∀ I : ideal P, a ∈ I :=
 ⟨λ _, by exactI ideal_Inter_nonempty.exists_all_mem, ideal_Inter_nonempty_of_exists_all_mem⟩
 
+end has_le
+
+section preorder
+variables [preorder P] {I J : ideal P} {x y : P}
+
+/-- The smallest ideal containing a given element. -/
+def principal (p : P) : ideal P :=
+{ carrier   := { x | x ≤ p },
+  nonempty  := ⟨p, le_rfl⟩,
+  directed  := λ x hx y hy, ⟨p, le_rfl, hx, hy⟩,
+  mem_of_le := λ x y hxy hy, le_trans hxy hy, }
+
+instance [inhabited P] : inhabited (ideal P) := ⟨ideal.principal default⟩
+
+@[simp] lemma principal_le_iff : principal x ≤ I ↔ x ∈ I :=
+⟨λ (h : ∀ {y}, y ≤ x → y ∈ I), h (le_refl x),
+ λ h_mem y (h_le : y ≤ x), I.mem_of_le h_le h_mem⟩
+
+@[simp] lemma mem_principal : x ∈ principal y ↔ x ≤ y := iff.rfl
+
+lemma mem_compl_of_ge {x y : P} : x ≤ y → x ∈ (I : set P)ᶜ → y ∈ (I : set P)ᶜ :=
+λ h, mt (I.mem_of_le h)
+
+lemma inter_nonempty [is_directed P (swap (≤))] (I J : ideal P) : (I ∩ J : set P).nonempty :=
```

Comment:
Does this need the `preorder` assumption?

Reply:
Nope, well spotted!